src/quotamodel.c

```diff
@@ -1515,10 +1532,31 @@ do_load_quotas(void)
 		}
 		else
 		{
-			update_limit_for_quota(quota_limit_mb * (1 << 20), segratio, quotaType, (Oid[]){targetOid, spcOid});
+			if (SearchSysCacheExists1(TABLESPACEOID, ObjectIdGetDatum(spcOid)))
+				update_limit_for_quota(quota_limit_mb * (1 << 20), segratio, quotaType, (Oid[]){targetOid, spcOid});
+			else
+				cleanConfigTables = true;
 		}
 	}
 
+	if (cleanConfigTables)
+	{
+		SPI_execute(
+		        "delete from diskquota.target"
+		        " where (quotaType = 2 and primaryOid not in (select oid from pg_namespace))"
+		        "    or (quotaType = 3 and primaryOid not in (select oid from pg_roles))"
+		        "    or tablespaceOid not in (select oid from pg_tablespace)",
+		        false, 0);
+		SPI_execute(
+		        "delete from diskquota.quota_config"
+		        " where (quotaType = 0 and targetOid not in (select oid from pg_namespace))"
+		        "    or (quotaType = 1 and targetOid not in (select oid from pg_roles))"
+		        "    or (quotaType = 4 and targetOid not in (select oid from pg_tablespace))"
+		        "    or (quotaType in (2, 3) and (targetOid, quotaType) not in (select rowId, quotaType from "
+		        "diskquota.target))",
+		        false, 0);
```

Comment:
Why is this not enough and should it also be done in the hook?

Reply:
"Cleaning at the loading time only is not enough, because scheme, role or tablespace can be dropped and the same object with the same oid can be created in the period between two loadings." - it is written in the PR description